[DO NOT MERGE] ostree: make selinux optional

Silvanoc opened the pull request:
Enable building with or without SELinux support via the go build tag
'selinux'.

Building with the build tag 'selinux' requires the presence of
'libselinux' to run, no matter if SELinux is enabled or not.

Building without the build tag 'selinux' eliminates the requirement of
having 'libselinux'.

This PR implements the initial intention of #546.

Closes #432.

Signed-off-by: Silvano Cirujano Cuesta <silvano.cirujano-cuesta@siemens.com>

Silvanoc commented:
This PR implements the "buildtags" approach described [here](https://github.com/containers/image/pull/546#issuecomment-454034245).

rhatdan commented on ostree/ostree_selinux.go:
```diff
@@ -0,0 +1,74 @@
+	var selinuxHnd *C.struct_selabel_handle
+
+	if os.Getuid() == 0 && selinux.GetEnabled() {
+		selinuxHnd, err := C.selabel_open(C.SELABEL_CTX_FILE, nil, 0)
```
This SELinux handle will not reset the one below.
Remove ':' from :=

Silvanoc replied:
@rhatdan thanks for the prompt review!

The error is not in the assignment, but in [line 29](https://github.com/containers/image/pull/562#discussion-diff-248706597R29), which isn't needed.

Removing it.

rhatdan commented:
@giuseppe @cgwalters PTAL

Silvanoc commented:
@mtrmac @vrothberg FYI

Silvanoc commented on ostree/ostree_dest.go:
```diff
@@ -430,16 +402,12 @@ func (d *ostreeImageDestination) Commit(ctx context.Context) error {
-
-		defer C.selabel_close(selinuxHnd)
+	mac := new(mandatoryAccessControl)
+	err = mac.Open()
```
It was my first approach, but I probably failed due to lack of Go expertise.

How to return the initialized object if everything is fine, but an error otherwise?

mtrmac replied:
Typically, such functions return a pointer, and
```go
if failed {
   return nil, someError
}
return &mandatoryAccessControl{
   // initialized members
}
```

or, perhaps just
```go
if failed {
   return mandatoryAccessControl{}, someError
}
return mandatoryAccessControl{
   // initialized members
}
```

Assuming that the callers do check the error return value, both approaches are equivalent in effect (and the performance impact is not really worth worrying about).

Silvanoc replied:
I strongly modified the code and this issue should be fixed by now.

I hope I got a more Go idiomatic implementation.

Silvanoc commented on ostree/ostree_dest.go:
```diff
@@ -194,33 +188,11 @@ func fixFiles(selinuxHnd *C.struct_selabel_handle, root string, dir string, user
-		if selinuxHnd != nil {
-			relPath, err := filepath.Rel(root, fullpath)
+		if mac.handler != nil {
+			err = mac.checkPermissions(root, fullpath, info.Mode())
```
You are right, fixing.

Silvanoc replied:
I strongly modified the code and this issue should be fixed by now.

giuseppe commented:
at the moment it is used only for atomic system containers and that will probably be dropped in future once we stop supporting them.  So before we put more effort on it, is there any real use case for this?  For what are you using the ostree backend?

Silvanoc commented:
> at the moment it is used only for atomic system containers and that will probably be dropped in future once we stop supporting them. So before we put more effort on it, is there any real use case for this? For what are you using the ostree backend?

@giuseppe I don't understand what is meant with "only for atomic system containers". I mean c/image is a library being used in "skopeo", "buildah", "libpod", "CRI-O". All of them might use this feature in the future right? All of them mostly RedHad Projects, but some other projects unknown to you might also use it.

Are you really thinking on dropping this feature? Don't you really see any other usage for it?

Let me try to describe you use-cases that are probably close to your domain (and my personal one) and then use-cases close to that of my company (embedded systems). I hope that I can convince you this way to keep and even promote OStree support on the mentioned tools.

Many of the use-cases being mentioned below can be addressed also with BtrFS, but there are enough scenarios where BtrFS is not desirable (mostly on embedded systems).

## Storage footprint reduction

It applies to all three domains (cloud, desktop and embedded).

Almost all current container solutions rely on layering (based on OverlayFS) for deduplication. It is much better than no deduplication at all, but file-based deduplication is even better.

You can observe the impact especially when having multiple images derived from different releases of the same Linux distribution (e.g. `debian:*`, `debian:*-slim`, `golang`, `buildpack-deps`, ...).

## Network usage reduction

Due to the above mentioned storage footprint reduction, using OStree for files provisioning combined with SPDY/HTTP2.0 in the above mentioned scenario (multiple images derived from different releases of the same Linux distribution) can strongly reduce the amount of data being downloaded.

This use-case is probably not relevant on cloud environments, but on embedded systems that might rely on mobile networks is relevant.

## Memory footprint reduction

This use-case is mostly relevant for embedded systems (where memory is typically a scarce resource).

Having the same library in two different layers implies loading the corresponding pages twice to memory (unless you have KSM or similar, which comes at a CPU price). The file-based deduplication of those two instances of the library would also provide automatic memory page deduplication (due to the fact that they have the same inode).

## Filesystem support

Having file-based deduplication with OStree for read-only files implies that no layering is needed at all. Certain of the mechanisms being implemented for containers are already resolved at filesystem level. For example, releasing the inodes of a file because they are not linked to any file is a mature battle-proof mechanism on Linux filesystems, cleaning up unused layers is still a pain in Docker.

This use-case would be helpful on all three domains.

## Avoid OverlayFS limitations

OverlayFS has a couple of limitations which don't apply to OStree that might be relevant in certain scenarios: overlay on overlay doesn't work, [non-supported system calls](https://docs.docker.com/storage/storagedriver/#limitations-on-overlayfs-compatibility), no garbage collection possible for whiteouts.

This use-case if probably more relevant on embedded systems, but not only.

We are still trying to find a solution for following problem: merging pre-provisioned images on a read-only partition with other images from another partition. We tried with overlay on overlay (what according to the documentation should work), but it doesn't work. Using OStree for the pre-provisioned images and laying the images of the other partition over it would resolve the problem.

giuseppe commented:
@Silvanoc the ostree backend in containers/image doesn't handle most of that.  It is a very specific implementation where most of the logic is handled by the atomic tool (https://github.com/projectatomic/atomic/blob/master/Atomic/syscontainers.py). Without that, the ostree backend in containers/image is not really usable.

For the deduplication part, you'd probably want to investigate the ostree deduplication support in containers/storage that works on top of overlay.  Even that has some problems though, and we will probably replace it with reflinks.

Silvanoc commented on ostree/ostree_selinux.go:
```diff
@@ -0,0 +1,72 @@
+// +build !containers_image_ostree_stub,selinux
```
I strongly modified the code and this issue should be fixed by now.

I have changed the buildtags to keep the current default: SELinux support.

I just replaced the buildtag `selinux` with `mac_stub` with the inverse logic.

Silvanoc commented on .travis.yml:
```diff
@@ -15,6 +15,7 @@ before_install:
 env:
   - BUILDTAGS='btrfs_noversion libdm_no_deferred_remove'
   - BUILDTAGS='btrfs_noversion libdm_no_deferred_remove containers_image_openpgp'
+  - BUILDTAGS='btrfs_noversion libdm_no_deferred_remove selinux'
```
According to [another comment](https://github.com/containers/image/pull/562#discussion_r248976216) SELinux is active by default and that's the build setup being tested by Travis. I have to have a deeper look at the possibility of adding another build setup to test the build without SELinux.

Silvanoc commented on ostree/ostree_dest.go:
```diff
@@ -194,33 +188,11 @@ func fixFiles(selinuxHnd *C.struct_selabel_handle, root string, dir string, user
 
-		if selinuxHnd != nil {
-			relPath, err := filepath.Rel(root, fullpath)
+		if mac.handler != nil {
```
You are right.

I strongly modified the code and this issue should be fixed by now.

Silvanoc commented on ostree/ostree_noselinux.go:
```diff
@@ -0,0 +1,22 @@
+	handler *unsafe.Pointer
+}
+
+func (m mandatoryAccessControl) Open() error {
```
Should be fixed by now.

giuseppe commented:
> > 
> 
> The PR is about enabling the possibility of having OStree in tools integrating this library (like Skopeo) without requiring `libselinux` and without breaking your working tools.

I am still not sure how you are going to use it, I was just trying to warn you that the "ostree" backend is not what you probably expect as it is used only for a very specific use case, that is going to be dropped in future.

In any case, code LGTM

giuseppe commented:
> @giuseppe Could we use this for saving space in the containers/storage world? Right now if I pull down 50 images, would ostree be able to benefit me, or would I be better off playing with reflinks?

we have an ostree deduplication in containers/storage for that.  It is still useful when reflinks are not usable, the implementation in containers/storage is completely unrelated with the backend in containers/image that is used to store images into an OSTree repository.

Silvanoc commented:
> I am still not sure how you are going to use it, I was just trying to warn you that the "ostree" backend is not what you probably expect as it is used only for a very specific use case, that is going to be dropped in future.

I don't really want to use the OStree backend of *c/image* for fetching from an OStree repository. At least not now. I'm rather interested on the deduplication of *c/storage*.

Due to the inclusion of *c/image* in *libpod*, *crio*, *skopeo* and *buildah* with its current dependency on `libselinux`, there is no way in any of them to use OStree (even if only interested in the *c/storage* part) without having `libselinux` available in the system.

Thanks @giuseppe for the warning!

giuseppe commented:
> The current ArchLinux package [announced a couple of days ago](https://twitter.com/Barthalion/status/1088164494851887105) by @barthalion shows the first public use-case for this patch, since it would eliminate the need for [patching](https://git.archlinux.org/svntogit/community.git/tree/trunk/0001-Revert-SELinux-support.patch?h=packages/podman) that the current [packaging shows](https://git.archlinux.org/svntogit/community.git/tree/trunk?h=packages/podman).
> 
> @giuseppe this is not exactly one of the main use-cases that I've mentioned in this PR, but at least one that I can easily reference grimacing

have you found any issue using `TAGS=containers_image_ostree_stub` ?  That should not disable the ostree support in `containers/storage`

Silvanoc commented:
I'm wondering if something like following in [containers/libpod/hack/ostree_tag.sh](https://github.com/containers/libpod/blob/c9e1f365e294135968bdd4d754dca7e1044587dc/hack/ostree_tag.sh) would do the trick for libpod...

```
#!/bin/bash
if ! pkg-config glib-2.0 gobject-2.0 ostree-1 2> /dev/null ; then
	echo containers_image_ostree_stub
else
	if ! pkg-config  libselinux 2> /dev/null ; then
		echo containers_image_ostree_stub ostree
	else
		echo ostree
	fi
fi
```

giuseppe commented:
we should probably just make it the default for containers/image and let skopeo override that.  To my knowledge there are no other programs using the ostree backend right now.

Silvanoc commented:
Probably @giuseppe hasn't considered that I wasn't using the build flag appropriately just because of lack of deep knowledge about the involved components and therefore he's been wondering which use-case can qualify for needing this patch and not using the right build flags combination.

In the end thanks to his insistency (that I have to admit was starting to irritate me a bit) I've seen the light. Apparently he's right and this patch is not needed, just the combination of build flags `ostree` (for having OStree support in containers/storage) and `containers_image_ostree_stub` (for not having OStree support in containers/image and this way no need for libselinux) appears to do the trick (at least in my initial tests).

The mentioned build flags combination `ostree containers_image_ostree_stub` appears to remove the need for having *libselinux* without disabling the usage of OStree for file-based deduplication. Exactly what I was aiming for with this PR. I just wan to run a couple of tests more before closing it.

I've created a [bug report](https://bugs.archlinux.org/task/61563?project=5&opened=29585) on ArchLinux to replace the patching with the right build flags. Please @barthalion have a look at it.

barthalion commented:
Thanks @Silvanoc, just got back from vacations today. I'll take a look at this today.

(And again, thanks for everyone involved for dilligence and patience.)

Silvanoc commented:
PR #587 is my third approach to this issue...

Hopefully and thanks to @giuseppe comments we have found this time a solution addressing all the concerns of @barthalion, @giuseppe, @vrothberg, @mtrmac and mine.

I'd nevertheless like to repeat, that I find the usage of OStree in c/storage can be a USP in certain scenarios.

I also find the usage of OStree in c/image meaningful for bandwidth improvement on fetching the images. At least until [this discussion](https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/icXssT3zQxE) leads to a standarized OCI-distribution solution.

mtrmac commented on ostree/ostree_selinux.go:
```diff
@@ -0,0 +1,78 @@
+// +build !containers_image_ostree_stub,!mac_stub
```
Please name the tag `containers_image_mac_stub`, this is a library and we don’t want to stomp on calling applications’ tags.

mtrmac commented on ostree/ostree_noselinux.go:
```diff
@@ -0,0 +1,14 @@
+type mandatoryAccessControl interface {
+	Close()
+	ChangeLabels(root string, fullpath string, fileMode os.FileMode) error
+}
```
Please move this type definition into a single location not gated by a build tag (`ostree_dest.go` would work), to ensure that the two definitions in `ostree_noselinux.go` and `ostree_selinux.go` can’t get out of sync.

And use private method names (lower-case `close` and `changeLabels`).

barthalion commented:
Finally found time to test it and I can confirm `containers_image_ostree_stub` makes it possible to build without SELinux. Sorry that it took so long…

rhatdan commented:
@Silvanoc What is the latest on this PR, is it still needed?

mtrmac commented:
Rereading parts of the various PRs/issues, if I understand things right, somewhere around https://github.com/containers/image/pull/587#issuecomment-471449534 it was affirmed that the `ostree:` backend is not useful for the contemplated use, and that it can be disabled without affecting the use of `ostree` mechanisms in c/storage — and has been disabled by default since.

Considering that the backend exists only for compatibility with `atomic` now, I don’t think it makes much sense to add features to that code.

@Silvanoc please reopen if I misunderstood the situation.
